### ISSUE TITLE
MoveScriptsToBottom: does not handle CDATA and comments correct

### ISSUE THREAD

**Issue by ssibitz** [primefaces/primefaces#2979]
I tried to use the new feature by setting primefaces.COLLECT_SCRIPTS to true.
After then and redeploy my application the primefacs focus and the DefaultCommand does not work anymore in Primefaces-6.2-SNAPSHOT:

I have the following login-mask created with primefaces looking like that in the original application.
As you can see i have also a focus on the username field:
![logon-showcase](https://user-images.githubusercontent.com/652355/33062638-3acb1334-cea0-11e7-88c3-5a717d869c37.png)

Here is the part of the view which is rendered as on the screenshot -
don't look on the controller part or on the design (I am using Bootstrap for beeing responsive) because only the JavaScript Part's are relevant.
This should be rendered for the marked components but they aren't rendered if i set COLLECT_SCRIPTS to true:
```
			<h:form id="logon-form" prependId="false" >
				<div>
					<div>
						<!-- Block with the logon-view -->			  	
					  	<div class="startup">
						  	<div class="LogonFormBlock" id="LogonInputContainer" >
						  		<center>
							  		<div class="row">
							  			<div class="col">
								  			<div class="col ml-auto">
                                                <h:panelGroup id="RenderGroup">
```
                                                   **<p:focus context="logon-form"/>**
```
                                                    <div class="row" >
                                                        <div class="col">
                                                            <p:messages id="msgPrivate" escape="false" autoUpdate="true" closable="false" showSummary="true" showDetail="true" styleClass="Percent100"/>
                                                        </div>
                                                    </div>
                                                    <div class="row">
                                                        <div class="col">
                                                            <div id="AutoGlow" class="LogonDescTitle noselect FloatLeft">#{tr.tr('Index.Label')}</div>
                                                        </div>
                                                    </div>
                                                    <!-- Login-Step -->
								  					<h:panelGroup id="LoginStep" rendered="#{loginControl.currentStep==EnumStep.STEP_LOGIN}">
                                                        <div class="row">
                                                            <div class="col">
                                                                <p:inputText id="UserName" autocomplete="off" value="#{loginControl.loginModel.user}" pt:placeholder="#{tr.tr('Index.UserName')}" styleClass="LogonInput" required="true" requiredMessage="#{tr.tr('Index.UserNameRequired')}"/>
                                                            </div>
                                                        </div>
                                                        <div class="row">
                                                            <div class="col">
                                                                <p:password id="Password" autocomplete="off" value="#{loginControl.loginModel.pwd}" pt:placeholder="#{tr.tr('Index.Password')}" styleClass="LogonInput" required="true" requiredMessage="#{tr.tr('Index.PasswordRequired')}"/>
                                                            </div>
                                                        </div>
								  					</h:panelGroup>
                                                    <!-- ClientSelection-Step -->
                                                    <h:panelGroup id="ClientStep" rendered="#{loginControl.currentStep==EnumStep.STEP_CLIENT}">
                                                        <div class="row">
                                                            <div class="col">
                                                                <p:inputText id="UserName2" disabled="true" readonly="true" value="#{loginControl.loginModel.user}" styleClass="LogonInput"/>
                                                            </div>
                                                        </div>
                                                        <div class="row">
                                                            <div class="col">
                                                                <p:inputText id="Password2" disabled="true" readonly="true" value="**********"  styleClass="LogonInput"/>
                                                            </div>
                                                        </div>
                                                        <div class="row">
                                                            <div class="col">
                                                                <p:selectOneMenu id="Client" value="#{loginControl.loginModel.client}" autoWidth="false" style="width: 100% !important;" styleClass="LogonInput" converter="omnifaces.SelectItemsIndexConverter" required="true" requiredMessage="#{tr.tr('Index.LocationRequired')}">
                                                                    <f:selectItem itemLabel="#{tr.tr('General.SelectOne')}" itemValue="#{null}" noSelectionOption="true"/>
                                                                    <f:selectItems value="#{loginControl.allClients}" var="client" itemLabel="#{client.label}" itemValue="#{client}"/>
                                                                </p:selectOneMenu>
                                                            </div>
                                                        </div>
                                                    </h:panelGroup>
                                                    <!-- General Button -->
                                                    <div class="row">
                                                        <div class="col">
                                                            <p:commandLink id="LoginButton" process="@this RenderGroup" update="RenderGroup msgPrivate" action="#{loginControl.Validate()}" styleClass="LogonButton noselect LogonButtonDetail Percent100" validateClient="true">
                                                                #{tr.tr('General.cmdContinue')}&nbsp;<i class="fa fa-sign-in IconRight"/>
                                                            </p:commandLink>
```
                                                            **<p:defaultCommand target="LoginButton"/>**
```
                                                        </div>
                                                    </div>
                                                    <div class="row">
                                                        <div class="col">
                                                            &nbsp;
                                                        </div>
                                                    </div>
                                                </h:panelGroup>
								  			</div>
							  			</div>
							  		</div>
								  	</center>
							  	</div>
						  	<div class="LogonFormFooter" id="LogonInputFooter">
						  		<center>
							  		<div class="row">
							  			<div class="col">
											<div class="row MarginTop20px">
												<div class="col">
										        	<p:link process="@this" update="@none" styleClass="LogonButton noselect LogonButtonLink Percent100" outcome="forgot">
										        		#{tr.tr('Index.Forgot')}
										        	</p:link>
												</div>
											</div>
										</div>
									</div>
								</center>
						  	</div>
					  	</div>
				  	</div>
				  </div>
			</h:form>

```

When i now set the COLLECT_SCRIPTS to true the marked components (in the body tag) are not rendered in the javascript-part, but when i set the flag to off i get the additional JavaScript and 
DefaultCommand / Focus will work again:

```
	<script type="text/javascript">
		$(function() {
			PrimeFaces.focus(null, "logon-form")
		});
	</script>
	<script id="j_idt63_s" type="text/javascript">
		$(function() {
			PrimeFaces.cw("DefaultCommand", "widget_j_idt63", {
				id: "j_idt63",
				target: "LoginButton"
			})
		});
	</script>

```
... so i think because this javascript is in the body-tag it will not be collected and so i miss it when i 
set the COLLECT_SCRIPTS to on.

**Comment by melloware** [comment]
Can you create a little reproducible test project so @tandraschko can debug.  Clone this repository https://github.com/primefaces/primefaces-test.git in order to reproduce your problem

**Comment by tandraschko** [comment]
Yep, otherwise i cant replicate... So please provide a smal

**Comment by tandraschko** [comment]
l as possible example

**Comment by ssibitz** [comment]
I thing i know now what the problem in detail is after debugging it:

I have used the **CDATA-TAG** in my inline-script's to be sure that it can used without any warnings
and can made a deploy without any errors.

_So i can reproduce the problem with a simpler example now - Here is the code of the view:_
```
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang="en"
      xmlns="http://www.w3.org/1999/xhtml"
      xmlns:f="http://java.sun.com/jsf/core"
      xmlns:h="http://java.sun.com/jsf/html"
      xmlns:p="http://primefaces.org/ui"
      xmlns:pt="http://xmlns.jcp.org/jsf/passthrough"
>
    <h:head>
        <title>Test</title>
    </h:head>
    <h:body>
        <h:outputScript>
            //<![CDATA[
                /* Dummy Example for using the CDATA-TAG */
                if (1==1 && 2==2) {
                    console.log("I am called !");
                }
            //]]>
        </h:outputScript>
        <h:form id="test-form" prependId="false">
            <p:focus context="test-form"/>
            <div class="ui-g">
                <p:messages id="msgPrivate" escape="false" autoUpdate="true" closable="false" showSummary="true" showDetail="true" style="width:100%;"/>
            </div>
            <div class="ui-g">
                <p:inputText id="UserName" autocomplete="off" value="#{usrInfo.echo}" pt:placeholder="Please enter your echo phrase:" style="width: 100%;"/>
            </div>
            <div class="ui-g">
                <p:commandLink id="TestButton" process="@this @form" update="msgPrivate UserName" action="#{usrInfo.ShowInfo()}">
                    Test&nbsp;<i class="fa fa-sign-in IconRight"/>
                </p:commandLink>
                <p:defaultCommand target="TestButton"/>
            </div>
        </h:form>
    </h:body>
</html>

```

... and here is the simple Controller:

```
@ViewScoped
@Named(value="usrInfo")
public class InfoController Serializable 
{
	private static final long serialVersionUID = 6527300364718959907L;
    // Test Datas
    private String echo;

    public String getEcho() {
        return echo==null?"Empty echo":echo;
    }

    public void setEcho(String echo) {
        this.echo = echo;
    }

    // Example Message
    public void ShowInfo()
    {
        FacesMessage msg = new FacesMessage(FacesMessage.SEVERITY_INFO,
                "This is a test : <br/>",
                getEcho());
        FacesContext.getCurrentInstance().addMessage(null, msg);
        FacesContext.getCurrentInstance().getExternalContext().getFlash().setKeepMessages(true);
    }
}

```

**When i set the COLLECT_SCRIPTS to true the following part is rendered out for the inline-script-part:**

```
<script type="text/javascript">
    var pf = window.PrimeFaces;
    if (pf) {
        pf.settings.locale = 'en_US';
    };
    var pf = window.PrimeFaces;
    //<![CDATA[
    /* Dummy Example */
    if (1 == 1 && 2 == 2) {
        console.log("I am called !");
    }
    //]]>;var pf=window.PrimeFaces;$(function(){PrimeFaces.focus(null, 'test-form');});var pf=window.PrimeFaces;pf.cw("InputText","widget_UserName",{id:"UserName"});var pf=window.PrimeFaces;$(function(){pf.cw("DefaultCommand","widget_j_idt12",{id:"j_idt12",target:"TestButton"});});var pf=window.PrimeFaces;OmniFaces.Unload.init('5753bf5f-36ed-435c-b490-0b96d8b96bf0');
</script>
```
Then the Focus and the DefaultCommand is no longer working...


When i turn the COLLECT_SCRIPTS to false the inline-scripts will be rendered like that:

```
        <script type="text/javascript">
            //<![CDATA[
                /* Dummy Example */
                if (1==1 && 2==2) {
                    console.log("I am called !");
                }
            //]]></script>
			<script type="text/javascript">$(function(){PrimeFaces.focus(null, 'test-form');});</script>
			<script id="UserName_s" type="text/javascript">PrimeFaces.cw("InputText","widget_UserName",{id:"UserName"});</script>
			<script id="j_idt12_s" type="text/javascript">$(function(){PrimeFaces.cw("DefaultCommand","widget_j_idt12",{id:"j_idt12",target:"TestButton"});});</script>
			<script type="text/javascript" src="/EcoCalcDD4Web/javax.faces.resource/unload.js.skf?ln=omnifaces&amp;v=2.6.5"></script>
			<script type="text/javascript">OmniFaces.Unload.init('912b8639-5286-4f03-9261-8f2a2c5cc17c')</script>
```
This example is working now because the CDATA-TAG is rendered "seperate" and so
the following code is not interpreted as comment's !

So i think one of the following should be done in that case:
* Render a CDATA-SCRIPT as a separate script
or
* Removing all CDATA-SCRIPT tag's inside the collected scripts...

**Comment by tandraschko** [comment]
I think we should remove all comments....
A PR is welcome, will merge it in 2,5 weeks ;)

**Comment by ssibitz** [comment]
I currently found a workaround for this by changing the CollectScriptsResponseWriter in a simple way:

1.) Removing all CDATA-TAG's from the Inline-Scripts (Start and End Tag's) in the minimizeInlineScript function
2.) Adding the CDATA-TAG on the Start and End of the Inline-Script's to be sure that the browser can interpret the script's in the right way.

So i changed the following method's of the CollectScriptsResponseWriter Class and then everything work's by now - as my current "Workaround":

```
    @Override
    public void endElement(String name) throws IOException {
        if ("script".equals(name)) {
            inScript = false;

            state.addInline(inline);
            state.addInclude(include);

            include.setLength(0);
            inline.setLength(0);
        }
        else if ("body".equals(name)) {
            for (int i = 0; i < state.getIncludes().size(); i++) {
                String src = state.getIncludes().get(i);
                if (src != null && !src.isEmpty()) {
                    getWrapped().startElement("script", null);
                    getWrapped().writeAttribute("type", "text/javascript", null);
                    getWrapped().writeAttribute("src", src, null);
                    getWrapped().endElement("script");
                }
            }

            getWrapped().startElement("script", null);
            getWrapped().writeAttribute("type", "text/javascript", null);
            // Start of the CDATA-TAG for Inline-Scripts
            getWrapped().write("//<![CDATA[\n");
            getWrapped().write(minimizeInlineScript(state.getInline()));
            // End of the CDATA-TAG for Inline-Scripts
            getWrapped().write("\n//]]>");
            getWrapped().endElement("script");

            getWrapped().endElement(name);
        }
        else {
            getWrapped().endElement(name);
        }
    }

    protected String minimizeInlineScript(StringBuilder script) {
        String minimized = script.toString();

        minimized = minimized.replace("PrimeFaces.settings", "pf.settings")
            .replace("PrimeFaces.cw", "pf.cw")
            .replace("PrimeFaces.ab", "pf.ab")
            .replace("window.PrimeFaces", "pf")
            .replace(";;", ";")
// Replacing all CDATA start/end tag's here:
            .replace("//<![CDATA[", "")
            .replace("//]]>", "")
            .replace("<![CDATA[","")
            .replace("]]>","");

        minimized = "var pf=window.PrimeFaces;" + minimized;

        return minimized;
    }

```

**Comment by ssibitz** [comment]
The only think the user has to keep in mind here is that all comments in any inline-scripts 
should be in a block-comment tag - then anythink is working fine 
as @tandraschko already has recognized...

I found a stackoverflow thread which is about this CDATA-TAG:
[https://stackoverflow.com/questions/66837/when-is-a-cdata-section-necessary-within-a-script-tag](https://stackoverflow.com/questions/66837/when-is-a-cdata-section-necessary-within-a-script-tag)

**Comment by tandraschko** [comment]
I think the problem could happen with all comments, so skipping all comments is the better solution and also minifizes a bit more

**Comment by ssibitz** [comment]
That's right - and if the COLLECT_SCRIPTS parameter is set to true
i think that the user will not need comment's in the JavaScript when Performance is the better way :-)

But i think using the CDATA-TAG only once can also be an improvement - or not ?

**Comment by ssibitz** [comment]
What i also found that when i am using the PrimeFaces-Extension **MasterDetail**,
the PrimePartial Rendering does not use the functionality of CollectScriptResponseWriter
so the eventually changed Script's are not on the end of the body...

Here a simple example which does not work because of the order of the CollectScript's:

```
<h:form id="FrmClients" prependId="false">
    <pe:masterDetail id="masterDetail" level="#{clients.currentLevel}">
		
		<!-- Overview of all clients -->
		<pe:masterDetailLevel level="1" id="LevelClients">  
			<h4>Client1</h4>
			<h4>Client2</h4>
                       <p:commandButton value="Go to Details">  
                          <pe:selectDetailLevel level="2"/>  
                     </p:commandButton>  
		</pe:masterDetailLevel>
		
		<!-- Details of a client -->
		<pe:masterDetailLevel level="2" id="LevelClientDetails">
			<p:commandLink id="LocationCancelButton" process="@form" update="@form" onclick="alert('Canceled !')">
				Cancel<i class="fa fa-chevron-left IconLeft"/>
			</p:commandLink>
			<p:hotkey bind="esc" handler="$('#LocationCancelButton').click();" />

			<p:commandLink id="LocationSaveButton" process="@form" update="@form" validateClient="true" onclick="alert('Saved !')">
				SaveDetails<i class="fa fa-floppy-o IconRight"/>
			</p:commandLink>
			<p:defaultCommand target="LocationSaveButton"/>
		</pe:masterDetailLevel>
    </pe:masterDetail>
</h:form>
```

When i press the "Go to Details" button i will produce an Ajax Partial Response,
but the writer here does not know anything about the already collected script's 
and so it renders the JavaScript-Part of the HotKey-Part directly after the LocationCancelButton button.

This will not work because all the Script's needed for execute the OnKey JavaScript are at the bottom of the body - so after the script...

**Comment by melloware** [comment]
@ssibitz I added a PR for removing the CDATA tags.  But I think you should open a new ticket for this issue above about Ajax updates including the scripts inline.

**Comment by ssibitz** [comment]
@melloware: 
Can you also remove the XHTML
CDATA-TAG's, because they can also be used by user:

`//<![CDATA[`

and

`//]]>`

?

Also at the begin of the Inline-Script-Tag the XHTML CDATA Tag's should be added at the beginning and the end of the inline script's - but don't forget to use a newline separator like that:

`getWrapped().write("//<![CDATA[\n");`

`getWrapped().write("\n//]]>");`

Otherwise inline script's containing as example :

`if (1==1 && 2==2)`

..will not work...

**Comment by tandraschko** [comment]
I dont think removing  cdata is good enoug, we should strip all comments, too. 

Also not sure about that ajax thing...
For example:
-the initial request renders a function "test" to the end of the body
-ajax requests does not collect anything (supporting this is tricky and the performance boost is around zero).
The ajax response will be evaluated on the client and if ir contains a call to "test", it should just work.

**Comment by tandraschko** [comment]
And the cdata and comment "ignoring" should be done inside the writer, not if everything is collected

**Comment by melloware** [comment]
I agree we should strip all comments too.  Let me take a shot at doing that.

**Comment by tandraschko** [comment]
Great. If the code becomes to complex and doesnt work for 100%, we should think about to NOT support this.
Complex inline scripts is a bad practice. At best inline scripts should call methods from script files. So you dont need cdata and comments in the inline script...

**Comment by melloware** [comment]
OK pushed another commit takes a shot at removing comment lines and CDATA elements as they are appended to the inline script.  Overall though I agree with @tandraschko inline scripts should be used minimally and for calling other functions inside JS.

If the user doesn't like the behavior they don't have to use COLLECT_SCRIPTS if they have heavy complex inline scripts.

**Comment by tandraschko** [comment]
I dont think that the PR is good enough to handle multiline and block comments. Also not sure if html script and h-outputScripts will be handelt the same. Overall im not sure if normal script tags are catched by our writer.
I think the only good way is to: check how both mojarra and myfaces calls the responsewriter for both html and jsf script tags and create unittests for all cases:
- hmlt script tag
- jsf script tag
- block comment
- normal comment
- new lines
- cdata
- facelet.skipcomments

**Comment by ssibitz** [comment]
Yes, that's right - and i know that ajax-request's don't supporting this because it's very tricky to handle this.

I think this only should be documented that when Activating the COLLECT_SCRIPTS
that all Dynamic Inline JavaScript's must be placed "outside" e.g. a MasterDetail Component
so that it can be rendered before...

e.g. for hotkey, ...

**Comment by tandraschko** [comment]
I dont understand that masterdetail problem but i can check it in 2,5 weeks - so please create a very small example + ticket.
Would also be good if you can rename this ticket to somethin like: collect_scripts doesnt handle comments and cdata correclty

**Comment by ssibitz** [comment]
I have changed the ticket name by now.

At the first i can say that a workaround is maybe that in general when collect_scripts  is activated
then on javascript-problems all thinks that are not working should be placed outside
Ajax-Component's like MasterDetail,
but i will create a separate Ticket...

**Comment by tandraschko** [comment]
this case should work now but i dont remove the comments for now. Thats fine for me, we didnt build a js minimizer here :D
@melloware could you also retest your application?

**Comment by melloware** [comment]
I have retested this morning and so far i haven't found any issues!

**Comment by tandraschko** [comment]
JFYI: will rename the feature to MOVE_SCRIPTS_TO_BOTTOM
